src/dataStore/DripsDataStore.sol

```diff
@@ -0,0 +1,240 @@
+
+/// @notice `DripsDataStore` is a helper contract allowing storing large `Drips` data structures
+/// on-chain, that would normally be kept off-chain, in events or in calldata.
+/// It can store lists of streams receivers, splits receivers and account metadata.
+/// Anybody can store lists, and once stored, a list can be loaded any number of times, by anybody.
+///
+/// `DripsDataStore` uses the SSTORE2 pattern, and it's cheaper than using the contract storage.
+/// For example storing a list of 100 streams receivers costs 1.4M gas, and loading about 32K.
+contract DripsDataStore {
+    /// @notice The hash representing an empty list.
+    bytes32 public constant EMPTY_HASH = 0;
+
+    /// @notice The streams receiver lists storage contract addresses.
+    // slither-disable-next-line naming-convention
+    mapping(bytes32 hash => address pointer) internal _streamsPointers;
+    /// @notice The splits receiver lists storage contract addresses.
+    // slither-disable-next-line naming-convention
+    mapping(bytes32 hash => address pointer) internal _splitsPointers;
+    /// @notice The account metadata lists storage contract addresses.
+    // slither-disable-next-line naming-convention
+    mapping(bytes32 hash => address pointer) internal _accountMetadataPointers;
+    /// @notice The streams history entries lists storage contract addresses.
+    // slither-disable-next-line naming-convention
+    mapping(bytes32 hash => address pointer) internal _streamsHistoryPointers;
+
+    /// @notice Emitted when a new streams receiver list is stored.
+    /// @param hash The hash of the stored streams.
+    event StreamsStored(bytes32 indexed hash);
+    /// @notice Emitted when a new splits receiver list is stored.
+    /// @param hash The hash of the stored splits.
+    event SplitsStored(bytes32 indexed hash);
+    /// @notice Emitted when a new account metadata list is stored.
+    /// @param hash The hash of the stored account metadata.
+    event AccountMetadataStored(bytes32 indexed hash);
+    /// @notice Emitted when a new streams history entries list is stored.
+    /// @param hash The hash of the stored streams.
+    event StreamsHistoryStored(bytes32 indexed hash);
+
+    /// @notice Store a list of streams receivers.
+    /// Storing an already stored list is valid, and it doesn't do anything.
+    /// @param streams The streams receivers list to store.
+    /// @return hash The hash of the stored list, that can be used to load it.
+    /// If `streams` is empty, it's `EMPTY_HASH`, which is `0`.
+    /// It's always equal to the streams hash in `Drips` of the stored list.
+    function storeStreams(StreamReceiver[] calldata streams) external returns (bytes32 hash) {
+        if (streams.length == 0) return EMPTY_HASH;
+        bytes memory data = abi.encode(streams);
+        hash = keccak256(data);
+        if (_streamsPointers[hash] == address(0)) {
```

Comment:
Okay, so basically we would allow a user to call the `storeStreams` twice with the same data.

I think it makes sense to not `revert` in this case. Why not allow it, we can skip the `_store` part. Theoretically, this could happen.

However, I would recommend to still emit the  `emit StreamsStored(hash);` in this case. 

In case someone is waiting for the `event`.

Reply:
Hmm, that's' a good point. OTOH the event has been emitted in the past, so the off-chain should be able to see it (unless they are working directly on a non-archival node on a forgetful network like Polygon). If we emit the event for double store, maybe we should also emit it with `EMPTY_HASH` when somebody calls `store...` for an empty array. Should we add an emit of each kind with `EMPTY_HASH` in the constructor too?